File: Data Structures & Algorithms/add-two-numbers/submission-7.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carry = 0
        head = ListNode(0)
        aux = head
        while l1 and l2:
            value = 0
            if l1 and l2:
                value = l1.val + l2.val
            elif l1:
                value = l1.val
            else:
                value = l2.val
            value += carry
            carry = value // 10
            value = value % 10
            head.next = ListNode(value)
            l1 = l1.next
            l2 = l2.next
            head = head.next
        while l1:
            value = l1.val
            value += carry 
            carry = value / 10
            value = value % 10
            head.next = ListNode(value)
            l1 = l1.next
            head = head.next
        while l2:
            value = l2.val
            value += carry 
            carry = value / 10
            value = value % 10
            head.next = ListNode(value)
            l2 = l2.next
            head = head.next
        if carry > 0:
            head.next = ListNode(carry)
        return aux.next         

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-7.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        auto prec{new ListNode{0, head}};
        auto it{head};
        head = prec;
        while (it) {
            auto c{it};
            int counter{1};
            while (counter < k && c) {
                c = c->next;
                counter ++;
            }
            if (c) {
                auto groupStart{it};
                auto groupEnd{c};
                auto cur{it};
                auto nextGroupStart{c->next};
                ListNode* prev{};
                while (cur != nextGroupStart) {
                    auto next{cur->next};
                    cur->next = prev;
                    prev = cur;
                    cur = next;
                }
                prec->next = groupEnd;
                groupStart->next = nextGroupStart;
                prec = groupStart;
                it = nextGroupStart;
            }
            else {
                prec->next = it;
                break;
            }
        } 
        return head->next;
    }
};
